### PR TITLE
Feature/graph geometry

### PR DESCRIPTION
**Pull request by michalk8**
Implementation of a graph geometry from the paper `Ground Metric Learning on Graphs`'

Sparse matrices are support via [host callbacks](https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html), wrapping `scikit-sparse`.

Limitations:
- not differentiable w.r.t. the edge weights (can be fixed via defining a custom JVP
  I'd include it either in a later PR, preferably with a better sparse Cholesky solver implementation (as a `jax` primitive, see below)

TODOs (pinging @marcocuturi for opinion):
- [x] decide whether we want to create experimental module, since we utilize some experimental features from `jax`, e.g., `experimentalhost_callback.call` and `experimental.sparse`
might also put `scipy/sklearn-sparse` as an extra requirement
**UPDATE**: added `scikit-sparse` to `experimental` requirements
- [ ] decide whether to directly wrap CHOLMOD as a jax primitive instead of doing the current (slightly hacky) way; this would require for us to create the C++ bindings, e.g., using pybind11 or update
**UPDATE**: will possibly done in the future
https://github.com/rgl-epfl/cholespy to work better with `jax`
- [x] Crank-Nicholson numerical scheme as in the original impl. [here](https://github.com/matthieuheitz/2020-JMIV-ground-metric-learning-graphs/blob/main_release/ml_kernels.py#L70)
**UPDATE**: yes

### CODE REVIEW

File: ott/core/sinkhorn.py [hunk 1]
```diff
@@ -326,7 +326,7 @@ class Sinkhorn:
     parallel_dual_updates: updates potentials or scalings in parallel if True,
       sequentially (in Gauss-Seidel fashion) if False.
     use_danskin: when ``True``, it is assumed the entropy regularized cost is
-      is evaluated using optimal potentials that are freezed, i.e. whose
+      is evaluated using optimal potentials that are frozen, i.e. whose
```
Comment (marcocuturi) [inline ott/core/sinkhorn.py:329]: nice catch :)

**Review by marcocuturi — COMMENTED** [review]
Thanks Michal! this looks awesome for now! 
a few remarks:
_- decide whether we want to create experimental module, since we utilize some experimental features from jax, e.g., experimentalhost_callback.call and experimental.sparse
might also put scipy/sklearn-sparse as an extra requirement_

this is a good question. at this moment we're only using experimental for this module. There's also checkify that might be relevant. I do not really have an opinion, what would be a best practice?

_- decide whether to directly wrap CHOLMOD as a jax primitive instead of doing the current (slightly hacky) way; this would require for us to create the C++ bindings, e.g., using pybind11 or update https://github.com/rgl-epfl/cholespy to work better with jax_

what's the problem with cholespy? 

_- Crank-Nicholson numerical scheme as in the original impl. [here](https://github.com/matthieuheitz/2020-JMIV-ground-metric-learning-graphs/blob/main_release/ml_kernels.py#L70)_

this would be a small modification, no? maybe worth indeed? any obstacle?

File: ott/core/decomposition.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+@jax.tree_util.register_pytree_node_class
+class CholeskySolver(abc.ABC, Generic[T]):
+  """Base class for Cholesky linear solver.
+
+  Args:
+    A: Symmetric positive definite matrix of shape ``[n, n]``.
+  """
+
+  _LOWER = True
```
[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/decomposition.py]: I've converted this to an argument + added a docstring.

File: ott/core/decomposition.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+class SparseCholeskySolver(
+    CholeskySolver[Union[jesp.CSR, jesp.CSC, jesp.COO, jesp.BCOO]]
+):
+  """Sparse Cholesky solver using :func:`jax.experimental.host_callback.call`.
+
+  Uses the CHOLMOD :cite:`cholmod:08` bindings from :mod:`sksparse`.
+
+  Args:
+    A: Symmetric positive definite matrix of shape ``[n, n]``.
+    beta: Decompose :math:`A + b * I` instead of :math:`A`.
```
[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/decomposition.py]: Done.

**Comment by michalk8** [comment]
> what's the problem with cholespy?

For now, I've just put `scikit-sparse` as optional requirement and import it only when needed (in the host callback function). This will likely create a very incomprehensible error message though, so I will explain in it better in the docs/find another way.

> what's the problem with cholespy?

Doesn't provide abstract evaluation for the tracer, nor the XLA translation rule (should be fine to add)
The problem with the abstract tracer is that we'd need to know the number of non-zeros the factor will have (so it would have to perform the symbolic factorization, if possible).
Otherwise, user has to provide `nse` (number of specified elements), for it to be jittable,
just like in [bcoo.fromdense](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.sparse.bcoo_fromdense.html).

> this would be a small modification, no? maybe worth indeed? any obstacle?

No obstacles as far as I can tell, will add!

File: ott/geometry/graph.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    t: Constant used when approximating the geodesic exponential kernel.
+      If `None`, use :math:`\frac{1}{|E|} \sum_{(u, v) \in E} weight(u, v)`
+      :cite:`crane:13`. In this case, the ``graph`` must be specified
+      and the edge weights are all assumed to be positive.
+    n_steps: Maximum number of steps used to approximate the heat kernel.
+    numerical_scheme: Numerical scheme used to solve the heat diffusion.
+    directed: Whether the ``graph`` is directed. If not, it will be made
+      undirected as :math:`G + G^T`. This parameter is ignored when  directly
+      passing the Laplacian, which is assumed to be symmetric.
+    tol: Relative tolerance with respect to the Hilbert metric, see
```
Comment (michalk8) [inline ott/geometry/graph.py:42]: @marcocuturi am not sure about this option, would consider removing it since it doesn't seem to help that much (at least in tests), but am happy to discuss.

Comment (marcocuturi) [reply ott/geometry/graph.py:42]: does it work better now with re-entering?

File: ott/geometry/geometry.py [hunk 5]
```diff
@@ -137,7 +137,8 @@ def cost_matrix(self) -> jnp.ndarray:
     if self._cost_matrix is None:
       # If no epsilon was passed on to the geometry, then assume it is one by
       # default.
-      cost = -jnp.log(self._kernel_matrix)
+      eps = jnp.finfo(self._kernel_matrix.dtype).tiny
```
Comment (marcocuturi) [inline ott/geometry/geometry.py:140]: smallest_normal ?

**Review by marcocuturi — APPROVED** [review]

File: ott/core/decomposition.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    Args:
+      A: Symmetric positive definite matrix of shape ``[n, n]``.
+      kwargs: Keyword arguments for the initialization.
+
+    Returns:
+      Sparse or dense Cholesky solver.
+    """
+    if isinstance(A, sp.spmatrix):
+      A = _scipy_sparse_to_jax(A)
+    if isinstance(A, (jesp.CSR, jesp.CSC, jesp.BCOO)):
```
[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/decomposition.py:80]: For flexiblity and less need for conversion, since for Cholesky, CSC representation is the best.

File: ott/geometry/geometry.py (hunk 5 above)
Comment (michalk8) [reply ott/geometry/geometry.py:140]: Doesn't seem to exist, both in `numpy` and in `jax`, despite the [docs](https://numpy.org/doc/stable/reference/generated/numpy.finfo.smallest_normal.html#numpy.finfo.smallest_normal) saying otherwise. Might need to open an issue to see why.

File: ott/geometry/graph.py (hunk 4 above)
Comment (michalk8) [reply ott/geometry/graph.py:42]: Don't know yet, still need to test it out.
